packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -42,105 +141,293 @@ export function createElement(original: parse5.Element): IRElement {
+}
 
-    // Walking up the AST and checking for each node to find if the identifer name is identical to
-    // an iteration variable.
-    for (let i = parentStack.length; i >= 0; i--) {
-        if (isElement(current)) {
-            const { forEach, forOf } = current;
+export function isBooleanLiteral(node: Expression | Literal): node is Literal<boolean> {
+    return node.type === LWCNodeType.Literal && typeof node.value === 'boolean';
+}
 
-            if (
-                forEach?.item.name === name ||
-                forEach?.index?.name === name ||
-                forOf?.iterator.name === name
-            ) {
-                return false;
-            }
-        }
+export function isForOf(node: BaseNode): node is ForOf {
+    return node.type === LWCNodeType.ForOf;
+}
 
-        current = parentStack[i - 1];
-    }
+export function isForEach(node: BaseNode): node is ForEach {
+    return node.type === LWCNodeType.ForEach;
+}
+
+export function isForBlock(node: BaseNode): node is ForBlock {
+    return isForOf(node) || isForEach(node);
+}
+
+export function isIfBlock(node: BaseNode): node is IfBlock {
+    return node.type === LWCNodeType.IfBlock;
+}
+
+export function isParentNode(node: BaseNode): node is ParentNode {
+    return isBaseElement(node) || isRoot(node) || isForBlock(node) || isIfBlock(node);
+}
+
+export function isProperty(node: BaseNode): boolean {
+    return node.type === LWCNodeType.Property;
+}
+
+export function isDynamicDirective(directive: Directive): directive is DynamicDirective {
+    return directive.name === LWCNodeType.Dynamic;
+}
+
+export function isDomDirective(directive: Directive): directive is DomDirective {
+    return directive.name === LWCNodeType.Dom;
+}
+
+export function isRenderModeDirective(directive: Directive): directive is RenderModeDirective {
+    return directive.name === LWCNodeType.RenderMode;
+}
+
+export function isPreserveComments(directive: Directive): directive is PreserveCommentsDirective {
+    return directive.name === LWCNodeType.PreserveComments;
+}
+
+export function isKeyDirective(directive: Directive): directive is KeyDirective {
+    return directive.name === LWCNodeType.Key;
+}
+
+export function getElementDirective(element: BaseElement, predicate: (dir: Directive) => boolean) {
+    return element.directives?.find((dir) => predicate(dir));
+}
+
+export function getDomDirective(element: BaseElement) {
+    return getElementDirective(element, isDomDirective) as DomDirective | undefined;
+}
+
+export function getKeyDirective(element: BaseElement) {
+    return getElementDirective(element, isKeyDirective) as KeyDirective | undefined;
+}
+
+export function getRootDirective(root: Root, predicate: (dir: Directive) => boolean) {
+    return root.directives?.find((dir) => predicate(dir));
+}
+
+export function getRenderModeDirective(root: Root) {
+    return getRootDirective(root, isRenderModeDirective)?.value.value as
+        | LWCDirectiveRenderMode
+        | undefined;
```

Comment:
I'm not sure if it was necessary to assert the return type as undefined as well.  I did it to be clear that there may not be a result.

Reply:
This function can be rewritten to avoid the type guards. 

```ts
return root.directives?.find(isRenderModeDirective)?.value.value;
```

I would also recommend adding return types for all the methods that are exported even if typescript does type inference. I would recommend adding return types to all the methods where the returned type value is not straightforward to infer. For instance, you want to ensure that the `getRenderModeDirective` does return `LWCDirectiveRenderMode` and not a standard string.